docs/testing.md

```diff
@@ -0,0 +1,140 @@
+        select(getUserById, { id: 5 }),
+        // A call mock, the second argument is the return value when this mock is used
+        call(fetchUserApi, { id: 5 }),
+        // A spawn mock, the second argument is the return value when this mock is used
+        spawn(trackUser),
+        // A run mock, the second argument is the return value when this mock is used
+        run(loginUser),
+      ])
+      // These are the expectations for the saga
+      // Expect the saga to call a function
+      // The arguments afterwards are what will be passed to the function
+      .toCall(fetchUserApi, 5, { all: true })
+      // Expect the saga to spawn a child bound function
+      .toSpawn(trackUser)
+      // Expect the saga to run a function
+      .toRun(loginUser, { id: 5 })
+      // Expect the saga to dispatch an action
+      .toDispatch(storeUser({ id: 5 }))
+      // Start the saga with the action
+      .dispatch(fetchUser(5))
+      // Do some assertions on the state
+      .toHaveFinalState()
+      // Start the whole saga and do the assertions
+      // You can provide a timeout as the first argument here for the saga to fail if it takes too long
+      // The default timeout is 10 seconds
+      .run()
+  );
+});
+```
+
+#### Mocks
+
+There are `select`, `call`, `run` and `spawn` mock helpers to easily setup mocks and have some type inference.
+These should only be used inside the tests and not in the actual saga code.
+
+```typescript
+import { select, call, run, spawn } from 'tsaga';
+```
+
+The test will also fail if you have unused mocks.
+This is an expected behavior so that the tests are always up to date with the source code and to keep the mocks as small as possible.
+
+#### Assertions
+
+The test will fail when there are assertions still left, but the saga has already finished.
+
+You also don't need to assert every single effect, you can easily skip an effect, but at the end of your test, every assertion needs to be used in the order you provided.
+
+Stack traces will show the location in the source code where the assertion failed.
+They will also show what is expected and what was received.
+
+##### Sub environments
+
+When you `run` or `spawn` a bound function, we will only forward the mocks to the child env.
+The child env will not use assertions as this can make tests hard to understand and assertions should only concern the saga which is currently tested.
+
+##### Take assertions
+
+Whenever you have a `take` in your saga, you will need to assert it to supply the right action which is returned to the saga.
+If you don't, the test will fail as there is no way to continue the saga without it.
+
+We are working on a solution for also triggering a timeout for a `take`.
+
+#### Assert on the final state
+
+Assertions on the final state are only possible when you provide a reducer.
+
+Only the provided top-level properties of the expected states are checked.
```

Comment:
Huh, when did this change? I would be opposed to that by default. In OMS it's rather easy to not test the full state, as we need some initial state anyway. For the final state we then to:

```ts
.toHaveFinalState({ ...initialState, basket: { total: 0 } })
```

That also at least ensure that a complete basket is tested and not just a partial one.

If one really wants this I would rather add a `map` parameter, so people can drill down into the part they want to assert one.

But the basic idea behind the testing was to cover the full state and ideally mock very little except network calls, such that a more realistic setup is tested, and not just what is convenient in the moment (but not as sure to find regressions over time).